sopel/test_tools.py

```diff
@@ -259,3 +272,257 @@ def run_example_tests(filename, tb='native', multithread=False, verbose=False):
         args.extend(["-n", str(cpu_count())])
 
     pytest.main(args)
+
+
+def botfactory(settings, autoloads=None):
+    """Build a test instance of Sopel, with a :class:`MockIRCBackend`.
+
+    :param settings: Sopel's configuration for testing purposes
+    :type settings: :class:`sopel.config.Config`
+    :param list autoloads: list of plugins to autoload after creating the bot
+    :return: a test instance of the bot
+    :rtype: :class:`sopel.bot.Sopel`
+
+    This will instanciate a :class:`~sopel.bot.Sopel` object, replace its
+    backend with a :class:`~MockIRCBackend`, and then autoload plugins. This
+    will automatically load the ``coretasks`` plugins, and every other plugins
+    from ``autoloads``::
+
+        bot = botfactory(settings, ['emoticons', 'find', 'remind'])
+    """
+    autoloads = set(autoloads or []) | {'coretasks'}
+
+    mockbot = Sopel(settings)
+    mockbot.backend = MockIRCBackend(mockbot)
+
+    usable_plugins = sopel.plugins.get_usable_plugins(settings)
+    for name in autoloads:
+        plugin = usable_plugins[name][0]
+        plugin.load()
+        plugin.register(mockbot)
+
+    return mockbot
+
+
+class IRCFactory(object):
+    """Factory to send IRC messages from a fake server to a test bot.
+
+    :param bot: test bot instance to send messages to
+    :type bot: :class:`sopel.bot.Sopel`
+
+    This factory helps developpers when they want to simulate an IRC server
+    sending messages to the bot.
+
+    You can generate a test bot instance with the :func:`botfactory` function::
+
+        >>> from sopel import test_tools
+        >>> bot = test_tools.botfactory(settings, ['emoticons'])
+        >>> irc = test_tools.IRCFactory(bot)
+        >>> irc.say(test_tools.UserFactory('TestSubject', '#test', '.shrug'))
+        >>> bot.backend.message_sent
+        [b'PRIVMSG #test :¯\\_(ツ)_/¯']
+
+    .. seealso::
+
+        This factory is best used with all the testing tools available:
+
+        * :func:`botfactory` to create a test instance of Sopel
+        * :class:`UserFactory` to fake IRC user who will join channels, send
```

Comment:
I _think_ this is what was intended.

```suggestion
        * :class:`UserFactory` to fake IRC users who will join channels, send
```